src/bin/migrate.ts

```diff
@@ -0,0 +1,295 @@
+
+  migrationState.bookmark = undefined
+  saveMigrationState(migrationState)
+}
+
+function readMigrationState(): MigrationState {
+  const migrationStateFileContent = fs.readFileSync(MIGRATION_STATE_FILE, {
+    encoding: 'utf8'
+  })
+  return asMigrationState(migrationStateFileContent)
+}
+
+function saveMigrationState(migrationState: MigrationState): void {
+  fs.writeFileSync(MIGRATION_STATE_FILE, wasMigrationState(migrationState), {
+    encoding: 'utf8'
+  })
+}
+
+function getDocumentIdentifiers(
+  documentId: string
+): { appId: string; partnerId: string } {
+  const parts = documentId.split(':')[0].split('_')
+  if (parts.length === 0) {
+    throw new Error(`Invalid documentId ${documentId}`)
+  }
+  const partnerId = parts.pop() as string
+  const appId = parts.join('_')
+  return { appId, partnerId }
+}
+
+/**
+ * Adds the createTime field to documents that are missing it.
+ * This is a required field for migration processing.
+ */
+async function initializeDocument(
+  docs: Array<StandardTx & nano.Document>
+): Promise<DbTx[]> {
+  const newDocs: DbTx[] = []
+  for (const doc of docs) {
+    const { partnerId } = getDocumentIdentifiers(doc._id)
+    const processor = processors[partnerId]
+
+    if (processor == null) {
+      // Add createTime to document for minimal requirement for
+      // migration/processing.
+      newDocs.push({ ...doc, createTime: new Date() })
+      datelog(`Not found ${partnerId} for document ${doc._id}`)
+      continue
+    }
+
+    if (doc.rawTx == null) {
+      // Add createTime to document for minimal requirement for
+      // migration/processing.
+      newDocs.push({ ...doc, createTime: new Date() })
+      datelog(`Missing rawTx for document ${doc._id}`)
+      continue
+    }
+
+    let standardTx
+    try {
+      standardTx = processor(doc.rawTx)
+    } catch (error) {
+      // Add createTime to document for minimal requirement for
+      // migration/processing.
+      newDocs.push({ ...doc, createTime: new Date() })
+      datelog(`Error processing ${doc._id}`, error)
+      continue
+    }
+
+    newDocs.push(
+      wasDbTx({
+        _id: doc._id,
+        _rev: doc._rev,
+        ...standardTx
+      })
+    )
+  }
+
+  return newDocs
+}
+
+/**
+ * Updates the documents with any new fields using its processor function.
+ */
+async function updateDocument(
+  docs: Array<StandardTx & nano.Document>
+): Promise<DbTx[]> {
+  const newDocs: DbTx[] = []
+  for (const doc of docs) {
+    const { partnerId } = getDocumentIdentifiers(doc._id)
+    const processor = processors[partnerId]
+
+    if (processor == null) {
+      datelog(`Not found ${partnerId} for document ${doc._id}`)
+      continue
+    }
+
+    if (doc.rawTx == null) {
+      datelog(`Missing rawTx for document ${doc._id}`)
+      continue
+    }
+
+    let standardTx
+    try {
+      standardTx = processor(doc.rawTx)
+    } catch (error) {
+      datelog(`Error processing ${doc._id}`, error)
```

Comment:
This should throw otherwise this document won't be migrated and we'd need to fix it before continuing.